subsys/bluetooth/controller/ticker/ticker.c

```diff
@@ -1067,24 +1112,57 @@ void ticker_worker(void *param)
 		/* Scheduled timeout is acknowledged to be complete */
 		ticker->ack--;
 
-		if (ticker->timeout_func) {
+		if (ticker->timeout_func
+#if defined(CONFIG_BT_TICKER_EXT)
+		    || (ticker->ext_data && ticker->ext_data->ext_timeout_func)
+#endif /* CONFIG_BT_TICKER_EXT */
+		) {
 			uint32_t ticks_at_expire;
 
 			ticks_at_expire = (instance->ticks_current +
 					   ticks_expired -
 					   ticker->ticks_to_expire_minus) &
 					   HAL_TICKER_CNTR_MASK;
 
+#if defined(CONFIG_BT_TICKER_EXT)
+#endif /* CONFIG_BT_TICKER_EXT */
 			DEBUG_TICKER_TASK(1);
 			/* Invoke the timeout callback */
-			ticker->timeout_func(ticks_at_expire,
+
+#if defined(CONFIG_BT_TICKER_EXT)
+			if (ticker->ext_data && ticker->ext_data->ext_timeout_func) {
+				struct ticker_expire_info other_expire_info;
+				struct ticker_ext_timeout_data data;
+				bool found = false;
+
+				if (ticker->ext_data->expire_info_id != TICKER_NULL) {
+					found = ticker_get_expire_info(instance,
+							ticker->ext_data->expire_info_id,
+							ticker->next, &other_expire_info);
+				}
+
+				data.ticks_at_expire = ticks_at_expire;
+				data.ticks_drift = ticks_drift;
+				data.remainder = ticker->remainder_current;
+				data.lazy = must_expire_skip ?
+					    TICKER_LAZY_MUST_EXPIRE :
+					    ticker->lazy_current;
+				data.force = ticker->force;
+				data.other_expire_info = found ? &other_expire_info : NULL;
+
+				ticker->ext_data->ext_timeout_func(&data, ticker->context);
+			} else
+#endif /* CONFIG_BT_TICKER_EXT */
```

Comment:
I assume you mean defer to `ticker_job`, and obtain the information for the next timeout, then present/provide the information in `ticker_worker`? This could work if it's acceptable to get information which is not 100% up to date, but how it looked at the last `ticker_job` run. There would have to be code running for every invocation of `ticker_job`, which manipulates the nodes.

Reply:
> Due to the O(n) nature of `ticker_get_expire_info`, we should perform this query in the `ticker_job` when the head ticker is a ticker_ext with `expire_info_id`. The `ticker_expire_info` be a member in `ext_data`.

That would work as well, but as @mtpr-ot mentions it would mean that we will need to fire off re-calculations whenever something in the ticker changes (which is potentially expensive - I don't have a good enough overview to know how expensive in practice though). It would add more complexity for certain though.

The question is if the added overhead of this calculation is bad enough to warrant that; There is an upper bound limited by the number of valid ticker IDs, so we can be sure it won't grow beyond a certain point. In our case, the worst-case of having to search through all valid ticker IDs will give an overhead of ~26 microseconds - which should be reasonable. Of course, if you have a ticker with something like 255 valid ticker IDs it could be problematic (though you could adjust the EVENT_OVERHEAD_START_US value to compensate) - but is that realistic?

Reply:
> I assume you mean defer to ticker_job, and obtain the information for the next timeout, then present/provide the information in ticker_worker? This could work if it's acceptable to get information which is not 100% up to date, but how it looked at the last ticker_job run. There would have to be code running for every invocation of ticker_job, which manipulates the nodes.

The calculation will be done before the ticker compare is setup in the `ticker_job`, that is, only when the head in the ticker list changes to being the ticker that needs the expire info. In timeline, execution of `ticker_get_expire_info()` followed by call to `ext_timeout_func()` or `ticker_timeout_func ()` follow the same order with the advantage that `ticker_worker` consumes lower CPU time at its execution priority level.

> That would work as well, but as @mtpr-ot mentions it would mean that we will need to fire off re-calculations whenever something in the ticker changes (which is potentially expensive - I don't have a good enough overview to know how expensive in practice though). It would add more complexity for certain though.

In practice, head does not change until the ticker expire unless `ticker_start`, `ticker_update`, or `ticker_stop` (that change the head) is called between the compare being set and ticker expires.

Reply:
> The calculation will be done before the ticker compare is setup in the `ticker_job`, that is, only when the head in the ticker list changes to being the ticker that needs the expire info. In timeline, execution of `ticker_get_expire_info()` followed by call to `ext_timeout_func()` or `ticker_timeout_func ()` follow the same order with the advantage that `ticker_worker` consumes lower CPU time at its execution priority level.

The problem with only running the calculation for the head is that you can have several tickers expire within the same ticker_worker call. In most cases it will be fine to just have the information prepared in advance for the head; But in cases where two (or more) tickers expire in the same call.

We have a couple of ways we can deal with this:

1. Try to determine in advance which tickers could be expiring in the next call to ticker_worker instead of only looking at the head. I think we would need some kind of fudge factor here, since it will vary across hardware and load how close tickers will have to be to risk expiring in the same call.
2. Do the calculation for the head only, and in the unlikely case we end up expiring more than the head *and* it/they need expiration info, we run the calculation inside ticker_worker. This would still mean that we would sometimes have some delay before the ticker_cb is called though, so maybe it doesn't make sense
3. Always calculate expiration info for all tickers that requires it, not just the head. This is potentially very expensive though, since you can potentially have a lot of them